### PR TITLE
ethtool-util: several cleanups

### LINKED ISSUE #19625: `ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.`
On an Asus F2A85-M PRO running Debian sid/unstable with systemd 247 (247.3-5), systemd-udevd logs the messages below.

```
$ journalctl -o short-precise -u systemd-udevd.service -b
-- Journal begins at Mon 2021-05-10 15:48:05 CEST, ends at Sun 2021-05-16 11:39:40 CEST. --
May 16 11:39:03.094695 kodi systemd[1]: Started Rule-based Manager for Device Events and Files.
May 16 11:39:03.608824 kodi systemd-udevd[181]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
May 16 11:39:04.016808 kodi systemd-udevd[182]: Using default interface naming scheme 'v247'.
May 16 11:39:04.034608 kodi systemd-udevd[182]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
May 16 11:39:04.176704 kodi systemd-udevd[151]: Using default interface naming scheme 'v247'.
May 16 11:39:04.180926 kodi systemd-udevd[151]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
```

Is this a Linux kernel driver problem?

Edit:

From the output [`journalctl -o short-precise -b`](https://github.com/systemd/systemd/files/6520249/journal-o-short-precise-b.txt), the first udev message appears before the Linux driver is loaded.

```
[…]
May 21 06:31:32.712612 kodi systemd-udevd[163]: ethtool: autonegotiation is unset or enabled, the speed and duplex are not writable.
[…]
May 21 06:31:32.796376 kodi kernel: hub 10-0:1.0: USB hub found
May 21 06:31:32.796701 kodi kernel: r8169 0000:04:00.0 eth0: RTL8168f/8111f, 08:60:6e:74:7a:51, XID 480, IRQ 34
May 21 06:31:32.797079 kodi kernel: r8169 0000:04:00.0 eth0: jumbo features [frames: 9194 bytes, tx checksumming: ko]
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #19625.

### CODE REVIEW

**Review by bluca — APPROVED** [review]
Superficially looks fine to me, but not familiar with this area

**Comment by yuwata** [comment]
Ouch. LGTM.com is nice!

File: src/shared/ethtool-util.c [hunk 1]
```diff
@@ -299,6 +299,13 @@ int ethtool_get_permanent_macaddr(int *ethtool_fd, const char *ifname, struct et
         return 0;
 }
 
+#define UPDATE(dest, val, flag) \
+        {                       \
+                typeof(val) _v = (val);                \
+                flag = flag || dest == _v;             \
+                dest = _v;                             \
+        }
```
Comment (poettering) [inline src/shared/ethtool-util.c]: hard to grok what this does. can you add comment? and I think you mean dest != _v, not dest == _v, right?

Comment (poettering) [reply src/shared/ethtool-util.c]: hmm, i think "flag" is a misnomer here, it's a boolean (a flag to me is always the thing that makes up a bitmask..) hence maybe rename flag → changed.

maybe avoid the double eval of flag so:

```c
…
{
        typeof(val) _v = (val);
        if (dest != _v)
                changed = true;
        dest = _v;
}
```

Comment (poettering) [reply src/shared/ethtool-util.c]: maybe enclose the macro in `do {…} while(false)`

**Review by poettering — CHANGES_REQUESTED** [review]

File: src/shared/ethtool-util.c (hunk 1 above)
Comment (yuwata) [reply src/shared/ethtool-util.c]: > hard to grok what this does. can you add comment? and I think you mean dest != _v, not dest == _v, right?

Ouch!

**Comment by yuwata** [comment]
@poettering Thank you for the comment. Now the macro is updated as suggested. PTAL.

**Comment by bluca** [comment]
Network tests look unhappy:

```
00:09:35 #0  0x00007f2a674a4d22 in raise () from /usr/lib/libc.so.6
00:09:35 #1  0x00007f2a6748e862 in abort () from /usr/lib/libc.so.6
00:09:35 #2  0x00007f2a677a0afc in log_assert_failed (text=text@entry=0x7f2a678730f9 "ethtool_fd", file=file@entry=0x7f2a67872d91 "src/shared/ethtool-util.c", line=line@entry=164, func=func@entry=0x7f2a67873b10 <__PRETTY_FUNCTION__.23> "ethtool_connect") at ../build/src/basic/log.c:862
00:09:35 #3  0x00007f2a676f6f92 in ethtool_connect (ethtool_fd=ethtool_fd@entry=0x0) at ../build/src/shared/ethtool-util.c:164
00:09:35 #4  0x00007f2a676f8223 in ethtool_get_permanent_macaddr (ethtool_fd=ethtool_fd@entry=0x0, ifname=ifname@entry=0x561eb64ca7c0 "vti6tun99", ret=ret@entry=0x561eb64ca850) at ../build/src/shared/ethtool-util.c:278
00:09:35 #5  0x0000561eb4ddc7c4 in decode_link (m=m@entry=0x561eb64c2480, info=0x561eb64ca7c0, patterns=patterns@entry=0x7ffd0d99b848, matched_patterns=matched_patterns@entry=0x561eb64ca7a0) at ../build/src/network/networkctl.c:559
00:09:35 #6  0x0000561eb4ddd95a in acquire_link_info (bus=<optimized out>, rtnl=<optimized out>, patterns=patterns@entry=0x7ffd0d99b848, ret=ret@entry=0x7ffd0d99b650) at ../build/src/network/networkctl.c:748
00:09:35 #7  0x0000561eb4de4bbf in link_status (argc=2, argv=0x7ffd0d99b840, userdata=<optimized out>) at ../build/src/network/networkctl.c:2415
00:09:35 #8  0x00007f2a67768a31 in dispatch_verb (argc=argc@entry=5, argv=0x7ffd0d99b840, argv@entry=0x7ffd0d99b828, verbs=verbs@entry=0x561eb4e030e0 <verbs>, userdata=userdata@entry=0x0) at ../build/src/shared/verbs.c:103
00:09:35 #9  0x0000561eb4dd871a in networkctl_main (argc=argc@entry=5, argv=argv@entry=0x7ffd0d99b828) at ../build/src/network/networkctl.c:2993
00:09:35 #10 0x0000561eb4dd874a in run (argc=5, argv=0x7ffd0d99b828) at ../build/src/network/networkctl.c:3015
00:09:35 #11 0x0000561eb4de4d82 in main (argc=<optimized out>, argv=<optimized out>) at ../build/src/network/networkctl.c:3018
```

**Comment by yuwata** [comment]
Ouch. Fixed.

**Pull request merged**